File: maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.max_count = -1

        def dfs(node, count):
            if node is None:
                return
            if count > self.max_count:
                self.max_count = count
            if node.left is not None:
                dfs(node.left, count + 1)
            if node.right is not None:
                dfs(node.right, count + 1)
        dfs(root, 0)
        return self.max_count + 1

File: algorithms/easy/maximum-depth-of-binary-tree.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


struct TreeNode *newNode(int val, struct TreeNode *left, struct TreeNode *right) {
    struct TreeNode *node = malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = left;
    node->right = right;
    return node;
}

int maxDepth(struct TreeNode* root){
    if (root == NULL)
        return 0;
    int lm = maxDepth(root->left) + 1;
    int rm = maxDepth(root->right) + 1;
    return lm > rm? lm: rm;
}

int main() {
    struct TreeNode *n15 = newNode(15, NULL, NULL);
    struct TreeNode *n7 = newNode(7, NULL, NULL);
    struct TreeNode *n20 = newNode(20, n15, n7);
    struct TreeNode *n9 = newNode(9, NULL, NULL);
    struct TreeNode *n3 = newNode(3, n9, n20);
    printf("%d\n", maxDepth(n3));
}
